### PR TITLE
Reduce `get_token_frame_atoms` memory usage

### LINKED ISSUE #225: Inference OOM from `get_token_frame_atoms`
When running a collection of queries of increasing size, I'm hitting an OOM in `get_token_frame_atoms` with a 5,247 residue monomer (41,842 atoms). The OOM doesn't happen when this query is run on its own, so I think there must be some memory fragmentation going on (unfortunately `clear_cache_between_steps` didn't seem to help). But also this is just a pretty large materialization since the pairwise distance computation here is `torch.sum(eps + (x[..., None, :] - x[..., None, :, :]) ** 2, dim=-1) ** 0.5`, materializing an 3*N_atom^2 intermediate (I'm seeing 42GB of peak memory from this call alone)

Some potential fixes:

- I think we can use `torch.cdist` here. I believe by default this uses the identity `||a - b||^2 = ||a||^2 + ||b||^2 - 2 a.b` for more than 25 points (or always if you set `use_mm_for_euclid_dist`). This has some negative consequences for precision as it ends up subtracting large numbers to get a small result, but given that we then feed it into topk, this shouldn't have that big an effect. It's also about 4x faster (just for the distance computation), though I don't think that really matters here.
- The same identity can also be implemented manually with torch ops, but there's not really an advantage I can see.
- We could chunk this function as well. Gets bit-exact output, but somewhat more complex, though less so because we already have all the chunking machinery. No significant slowdown relative to current implementation, but massively reduced memory.
- Compute per coordinate separately and use a running sum. Same memory usage as cdist, without the catastrophic cancellation issues so much more precise results. Slightly faster than current implementation. Not bit-exact like chunking by atoms.

I had Claude compare the approaches. Tested by capturing snapshots of the tensors going into this function for the 14 inputs I used (49-4563 residues, 366-36,356 atoms), so we have realistic coordinates. But because distances are only used for frame atoms when is_atomized is true, I artificially flipped all of them to set them as atomized (would've been better to use actual atomized inputs, but this is what I had on hand). cdist changes the output of `phi` for four atoms total in the dataset (per input mean 0.05%, max 0.5%). None of the others change it at all and none of them change the `valid_frame_mask`. cdist and per-coord drop the memory requirement for 42k atoms from 42GB to 13GB and chunking in chunks of 1024 atoms drops it to 7GB. For any of those, something else in the function ends up dominating peak memory, which drops from 47GB to 33GB.

I think I would probably just use cdist here: it isn't a super precision-sensitive location and it's a simple fix.

Total function memory usage can be further reduced by converting `atom_mask` to bool and doing boolean operations with the masks. Then peak memory drops to 42GB with the current distance computation, 10GB with cdist and 15GB with chunking per coord or by atom.

There are some other locations that also do the distance calculation. I think these aren't run during inference, so I haven't hit them, but it might be worth investigating them as well. They don't get fed straight into a topk though, so there are more precision and differentiability concerns.

**Configuration:**
 - OS: Ubuntu 22.04
 - GPU used and memory available: MI300X 192GB
 - Python version (if relevant): 2.10
 - Installation path: pip installed from source

### PR DESCRIPTION
**Pull request by GMNGeoffrey**
**Summary**
I am hitting OOMs in `get_token_frame_atoms` for large monomer inputs around 5k residues (42K atoms). This function contributes ~47GB to the peak memory usage. We can bring that down to ~10GB by avoiding intermediate materializations and operating on boolean rather than fp32 masks.

**Changes**
- Replace manual pairwise distance computation with `torch.cdist`
- Convert masks to booleans
- Use in-place operations 

**Related Issues**
- Fixes https://github.com/aqlaboratory/openfold-3/issues/225

**Testing**
Verified outputs change only marginally across 14 monomer inputs with a range of sizes (49-4563 residues, 366-36,356 atoms). cdist changes the output of phi for four atoms total in the dataset (per input mean 0.05%, max 0.5%). The mask changes don't affect the output at all.

**Other Notes**
Using `torch.cdist` can create numerics issues with floating point math, as it computes a small number as the subtraction of two large ones. Since we are feeding the result directly into topk, I think this doesn't end up mattering too much. In my testing, there were only a few cases where the output differed at all.

### CODE REVIEW

**Comment by christinaflo** [comment]
And donot_use_mm_for_euclid_dist would remove the memory gains with this? I'm a little nervous since this is also used in training for the pae loss, I'd want a better idea of how much this changes outputs for a variety of different inputs. Separately, I was also thinking of moving this and other related indexing logic to the data pipeline and passing it in as a feature instead of computing it on the fly.

**Comment by GMNGeoffrey** [comment]
Yeah I think basically all the gains are from using matmul. One of the alternatives I mentioned in the linked issue is chunking here if you're worried about the effect on the outputs. I think feeding into topk here makes this not a huge concern, but totally understand if you're leary of this sort of change without more thorough testing
